### ISSUE TITLE
Setup multiple service accounts for WMAgent containers

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#11944]
**Impact of the new feature**
WMAgent containerized solution

**Is your feature request related to a problem? Please describe.**
Given that databases and the WMAgent container share the same mount points, we need to define a common service account that will have access/permissions to the base area.

**Describe the solution you'd like**
Define the expected service accounts (for FNAL, CERN and Tier0) in the CMSKubernetes dockerfiles for WMAgent, CouchDB and MariaDB; and identify which account to use during runtime.

A container which  is having the best solution (modulo the extra accounts' ids) is the [MariaDB container](https://github.com/dmwm/CMSKubernetes/blob/master/docker/pypi/wmagent-mariadb/Dockerfile). So we should use it as an example and adopt the same logic in  all of the rest.   

**Describe alternatives you've considered**
Not sure

**Additional context**
@todor-ivanov please add further information that I might be missing.

Sub-task for this meta issue: https://github.com/dmwm/WMCore/issues/11314

**Comment by vkuznet** [comment]
Alan, I need more information about this issue. What does it mean define service account. Does it mean to setup appropriate env variable with some name, in this case what is this env variable? Otherwise, which parts of code/services use service account and how? Also, I need to know what is full set of service accounts used by WM services. So far I'm aware that:
- CERN WMAgent uses cmst1
- FNAL WMAgent uses cmsdataops
- Tier0, I don't know

According to https://github.com/dmwm/WMCore/blob/master/deploy/deploy-wmagent.sh script it relies on `whoami` output to determine service account, and then use it in `scp` commands. But it is unclear to me where and how service account is used by other parts of WM eco-system.

Also, the descriptions says: ` identify which account to use during runtime`. I don't understand that we can change to an account at run time. We create a docker image with specific `USER`, e.g. `cmst1`. This user will be used when image is unpacked and all services will already run within this account which is defined by the time we create docker image. We can't change user to another user at run time. At run time we can identify the host name from infrastructure but I doubt we can switch to another user, and restart services under another account. We can create multiple Linux accounts during docker image manifest, and those will be created, but only one (which will be specified within Dockerfile) will be used to start services. If you want at runtime to switch to another Llinux account then our options will be only to use `sudo`, but it implies we'll need to restart all services which already would be started under default `USER` account.

Finally, in a description of this issue you wrote `Given that databases and the WMAgent container share the same mount points`. What does it mean that databases (which ones) and WMAgent container share mount points, which mount points?

In my view, if specific user must be used by services at runtime our only option to define such user name via env variable. But it means that services must use this env variable and I have no idea if it can be done. Another option is to create different docker images and pass user name at docker build as an option. Then, at a time we will install the image we'll specify which image to use and it will have proper account. This option will require changes to deployment script which will deal with docker images. But again I don't know if it is considered as an option for this issue. 

In order to proceed I suggest that you clarify all of the above nuances in a description of this issue.

**Comment by vkuznet** [comment]
My questions were answered via email thread exchange. In particular,
- the reason we need UID:GID is overlap of file IO operations between HTCondor and WM components, i.e. they write to the same area
- In order to do separation of this overlap will require changes both in HTCondor and WM codebase
- UID:GID can be passed via run-time as docker options

New PR using UID:GID is here: https://github.com/dmwm/CMSKubernetes/pull/1464

**Comment by vkuznet** [comment]
I would like to update this issue, even though we found a way to pass UID:GID to docker at run time they should match with existing /etc/passwd and /etc/groups entries. One solution was to pass them from a host which runs docker container. But we found that some account reside on AFS and therefore do not exist in local /etc files. Therefore it seems that until we create custom /etc/passwd and /etc/groups we still face similar issue reported earlier, i.e. we can't setup dynamically UID:GID unless we have their definition in a container. And, here we have two problems:
1. FNAL security team prohibits us from creating docker image with hard-coded UID:GID for security concerns
2. we can not pass UID:GID at run time unless their are already present in local /etc/passwd and /etc/groups, and on CERN nodes we have special accounts residing on AFS and not local /etc areas.

At the moment I can't find any solution, unless we puppetize creation of /etc area with our accounts, to satisfy both items shown above.

@amaltaro , @todor-ivanov please share your thoughts on this, otherwise I see that this issue in stalemate and cannot be resolved.

**Comment by amaltaro** [comment]
@vkuznet I think Todor has many other items in his pipeline, so I suggest we try to sort this one ourselves.
Let's try to experiment a bit with these options that we discussed today, understand what the actual error is and try to come up with a solution/alternatives to this problem. I will likely only be able to look into it tonight.

**Comment by todor-ivanov** [comment]
@amaltaro @vkuznet 

The solution seems to be not that  far away from where we are now. I can foresee two possible paths actually:
* We try to enable sssd module in the  CERN Containers  (trivial - just add one line in nsswitch), and map the /etc/passwd files for FNAL machines. This is actually acheivable 
* Creae the /etc/passwd file at runtime by merging the already existing `/etc/passwd` file on the host and current user details (regardless where they come from) fetched by using the command I mentioned above.  I can give it a try in few minutes and will tell you which one works better. (both are a single line implementations)

**Comment by vkuznet** [comment]
- Why do we need sssd module in CERN containers?
- where `getent` will read user's password?

My point is that the purpose of container should allow us to run it anywhere, and yet the proposed solutions will require a dedicated node (both at CERN and FNAL) in order to properly run a container. In this regard what do we gain of using a container if we still need to have dedicated nodes to run it? Performance wise container will still be slower that native VM, and if we still need native VM with specific account I would like to understand why do we need a containers then?

**Comment by todor-ivanov** [comment]
Hi @vkuznet 

I can answer  three out of your four questions. Here:

> Why do we need sssd module in CERN containers?

Because the user details come through sssd for CERN Nodes.  So there are two ways, we either read the the user details at runtime (the current user, who is starting the container), or we link the container to query the sssd as a  second user database once it fails to find the user details directly into the `/etc/passwd` files. This setup is really simple - we just need to cnfigure the `/etc/nsswitch.conf` file inside the container to query the user databsaes in the following order:
```
grep passwd /etc/nsswitch.conf  
passwd:     files sss
```

> where getent will read user's password?

As I said above - on the  host  (regardless if it is CERN or FNAL) getent always follows  the order for quering the user databases as configured in `/etc/nsswitch.conf`. So it will do its best to find the user details in the this case in the order I listed - first  `/etc/passwd` then `sssd`. And here comes the simplicity of the second solution I came up with: https://github.com/dmwm/CMSKubernetes/pull/1467/files#diff-d276e67ac0a9cb57c1f967acb04ff6f223d21bb3cded596c8de279941c934dcfR63-R69  - we do not care where and in what order the host running the container fetches the user details from. It will always match what we configure inside the container at runtime with what is at the current host. 


> My point is that the purpose of container should allow us to run it anywhere, and yet the proposed solutions will require a dedicated node (both at CERN and FNAL) in order to properly run a container.

We are still tied to those machines because of the HTCondor setup and the fact that we  cannot share any space between condor and the agent outside the node. We will get there, but not at this stage.
